File: 2025-2/sesion-de-clase-16/SoftProgPersistencia/DAOImpl/RRHH/EmpleadoDAOImpl.cs
using System;
using System.Data.Common;
using System.Data;
using PUCP.SoftProg.Modelo.RRHH;
using PUCP.SoftProg.Persistencia.DAO.RRHH;
using PUCP.SoftProg.Modelo;

namespace PUCP.SoftProg.Persistencia.DAOImpl.RRHH {
    public class EmpleadoDAOImpl : BaseDAOImpl<Empleado>, IEmpleadoDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, Empleado empleado) {

            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarEmpleado";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idArea", DbType.Int32, empleado.Area.Id);

            if (empleado.CuentaUsuario != null) {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, 
                    empleado.CuentaUsuario.Id);
            }
            else {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            }

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, empleado.Dni);

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, empleado.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, empleado.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, empleado.Genero);

            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, empleado.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_cargo", DbType.String, empleado.Cargo);
            this.AgregarParametroEntrada(cmd, "@p_sueldo", DbType.Double, empleado.Sueldo);

            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, empleado.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, Empleado empleado) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarEmpleado";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idArea", DbType.Int32, empleado.Area.Id);
            if (empleado.CuentaUsuario != null) {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32,
                    empleado.CuentaUsuario.Id);
            }
            else {
                this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, DBNull.Value);
            }
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, empleado.Dni);

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, empleado.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, empleado.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, empleado.Genero);

            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, empleado.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_cargo", DbType.String, empleado.Cargo);
            this.AgregarParametroEntrada(cmd, "@p_sueldo", DbType.Double, empleado.Sueldo);

            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, empleado.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, empleado.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarEmpleado";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarEmpleadoPorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarEmpleados";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override Empleado MapearModelo(DbDataReader reader) {
            return new Empleado {
                Id = Convert.ToInt32(reader["id"]),
                Dni = Convert.ToString(reader["dni"]), 
                Nombre = Convert.ToString(reader["nombre"]),
                ApellidoPaterno = Convert.ToString(reader["apellidoPaterno"]), 
                Genero = (Genero)Enum.Parse(typeof(Genero), Convert.ToString(reader["genero"])),
                FechaNacimiento = Convert.ToDateTime(reader["fechaNacimiento"]),
                Cargo = (Cargo)Enum.Parse(typeof(Cargo), Convert.ToString(reader["cargo"])), 
                Sueldo = Convert.ToDouble(reader["sueldo"]), 
                IsActive = (bool)reader["activo"],
                Area = new AreaDAOImpl().Leer(
                    Convert.ToInt32(reader["idArea"]))
            };
        }

        protected DbCommand ComandoBuscarPorDni(DbConnection conn, string dni) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarEmpleadoPorDni";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.Int32, dni);

            return cmd;
        }

        public Empleado BuscarPorDni(string dni) {
            return EjecutarComando(conn => {
                using (DbCommand cmd = this.ComandoBuscarPorDni(conn, dni)) 
                using (DbDataReader reader = cmd.ExecuteReader()) {
                    if (!reader.HasRows) {
                        Console.Error.WriteLine("No se encontro el empleado con dni: " + dni);
                        return default;
                    }

                    reader.Read();
                    return this.MapearModelo(reader);
                }
            });
        }
    }
}

File: 2026-1/sesion-de-clase-10/usa-procedimientos-almacenados/SoftProg/Referencia/softprog/daoimpl/rrhh/EmpleadoDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.rrhh;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.rrhh.EmpleadoDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.Genero;
import pe.edu.pucp.inf30.softprog.modelo.rrhh.Cargo;
import pe.edu.pucp.inf30.softprog.modelo.rrhh.Empleado;

/**
 *
 * @author eric
 */
public class EmpleadoDAOImpl extends BaseDAO<Empleado> implements EmpleadoDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, Empleado modelo) 
            throws SQLException {
        
        String sql = "{call insertarEmpleado(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);   
        cmd.setInt("p_idArea", modelo.getArea().getId());
        
        if (modelo.getCuentaUsuario() != null) {
            cmd.setInt("p_idCuentaUsuario", modelo.getCuentaUsuario().getId());
        }
        else {
            cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        }
        
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", 
                new Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_cargo", String.valueOf(modelo.getCargo()));
        cmd.setDouble("p_sueldo", modelo.getSueldo());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            Empleado modelo) throws SQLException {
        
        String sql = 
                "{call modificarEmpleado(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idArea", modelo.getArea().getId());
        
        if (modelo.getCuentaUsuario() != null) {
            cmd.setInt("p_idCuentaUsuario", modelo.getCuentaUsuario().getId());
        }
        else {
            cmd.setNull("p_idCuentaUsuario", Types.INTEGER);
        }
        
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", 
                new Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_cargo", String.valueOf(modelo.getCargo()));
        cmd.setDouble("p_sueldo", modelo.getSueldo());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) 
            throws SQLException {
        
        String sql = "{call eliminarEmpleado(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) 
            throws SQLException {
        
        String sql = "{call buscarEmpleadoPorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) 
            throws SQLException {
        
        String sql = "{call listarEmpleados()}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected Empleado mapearModelo(ResultSet rs) throws SQLException {
        Empleado empleado = new Empleado();
        empleado.setId(rs.getInt("id"));
        empleado.setArea(new AreaDAOImpl().leer(rs.getInt("idArea")));
        
        int idCuentaUsuario = rs.getInt("idCuentaUsuario");
        if (!rs.wasNull()) {
            empleado.setCuentaUsuario(
                    new CuentaUsuarioDAOImpl().leer(idCuentaUsuario));
        }
        empleado.setDni(rs.getString("dni"));
        empleado.setNombre(rs.getString("nombre"));
        empleado.setApellidoPaterno(rs.getString("apellidoPaterno"));
        empleado.setGenero(Genero.valueOf(rs.getString("genero")));
        empleado.setFechaNacimiento(rs.getTimestamp("fechaNacimiento"));
        empleado.setCargo(Cargo.valueOf(rs.getString("cargo")));
        empleado.setSueldo(rs.getDouble("sueldo"));
        empleado.setActivo(rs.getBoolean("activo"));

        return empleado;
    }

    protected PreparedStatement comandoBuscarPorDni(Connection conn, String dni)
        throws SQLException {
        String sql = "{call buscarEmpleadoPorDni(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_dni", dni);
        
        return cmd;
    }
    
    @Override
    public Empleado buscarPorDni(String dni) {
        return ejecutarComando(conn -> {
            try (PreparedStatement cmd = this.comandoBuscarPorDni(conn, dni)) {
                ResultSet rs = cmd.executeQuery();

                if (!rs.next()) {
                    System.err.println("No se encontro el registro con "
                            + "dni: " + dni);
                    return null;
                }

                return this.mapearModelo(rs);
            }
        });
    }
}
